scipy/stats/tests/test_stats.py

```diff
@@ -6704,6 +6707,154 @@ def test_dist_perm(self):
         assert_approx_equal(pvalue_dist, 0.001, significant=1)
 
 
+class TestFNI:
+    @pytest.mark.parametrize(("distname", "shapes"), distcont)
+    def test_FastNumericalInverse(self, distname, shapes):
+        slow_dists = {'ksone', 'kstwo', 'levy_stable', 'skewnorm'}
+        fail_dists = {'beta', 'gausshyper', 'geninvgauss', 'ncf', 'nct',
+                      'norminvgauss', 'genhyperbolic'}
+
+        if distname in slow_dists:
+            pytest.skip("FastNumericalInverse is not fast enough.")
+        if distname in fail_dists:
+            pytest.xfail("FastNumericalInverse fails; should fix.")
```

Comment:
This test fails for the distributions in the `fail_dists` set. Ultimately, all of these failures can be attributed to issues in the implementation of the distribution. This basic `FastNumericalInverse` algorithm is pretty robust, but it cannot be expected to be more accurate than the implementation of the distribution itself.

`genhyperbolic` fails because it doesn't meet the tolerances. During execution, we get these warnings:
```
C:\Users\matth\scipy\scipy\stats\_continuous_distns.py:3087: IntegrationWarning: The occurrence of roundoff error is detected, which prevents 
  the requested tolerance from being achieved.  The error may be 
  underestimated.
  t1 = integrate.quad(llc, -np.inf, x)[0]
C:\Users\matth\scipy\scipy\stats\_continuous_distns.py:3092: RuntimeWarning: Infinite values encountered in scipy.special.kve. Values replaced by NaN to avoid incorrect results.
  warnings.warn(msg, RuntimeWarning)
```
The resulting `FastNumericalInverse` is accurate for most points, but fails the accuracy test at some.

For the rest of the distributions, an error message is produced during the test:
```
ncf: `x` must be strictly increasing sequence.
nct: `x` must be strictly increasing sequence.
geninvgauss: `x` must be strictly increasing sequence.
beta: `x` must be strictly increasing sequence.
gausshyper: `dydx` must contain only finite values.
norminvgauss: Failed to converge after 100 iterations.
```

The ones that say "`x` must be strictly increasing sequence." fail because that error occurs when creating the Hermite spline. Ultimately, this signals an issue with the distribution - usually insufficient precision in the tails. For instance, in the case of `nct`:
```
ppf = stats.nct(14, 0.24045031331198066).ppf(1e-12/10)  # -22.68184321148624
cdf = stats.nct(14, 0.24045031331198066).cdf(ppf)  # exactly 0, but shouldn't be
isf = stats.nct(14, 0.24045031331198066).isf(1e-12/10)  # 22.68184332990839
cdf = stats.nct(14, 0.24045031331198066).cdf(isf)  # exactly 1.0, but shouldn't be
```
When the cdf is evaluated at other points in the grid, it gets exactly 0 or 1 again, so the inverse of SciPy's implementation of the CDF function is not a function. I can think of ways to get around this, but the time would be better spent fixing the distributions' `ppf` and `isf` methods rather than working around their deficiencies. 

`norminvgauss` fails because `norminvgauss(1.25, 0.5).isf(1e-12/10)` fails with that error message.

`gausshyper` fails because
```
shapes = (13.7637716041307, 3.118963664868143, 2.514598035018302, 5.1811649903971615)
stats.gausshyper(*shapes).isf(1e-12/10)  # 1.0
stats.gausshyper(*shapes).pdf(1.0)  # exactly 0
```
and then the derivative `1/f` that gets passed into the spline has an infinity. Again, this is really an issue with the `isf` method. It definitely shouldn't be returning the right end of the support (1.0) when called with an argument a few orders of magnitude greater than machine epsilon.

Most of these test pass if I pass in a tolerance explicitly when I generate the `FastNumericalInverse`, e.g.
```
stats.FastNumericalInverse(dist, tol=tol)
```
and relax the test tolerance accordingly. For a tolerance of 1e-8, all except `ncf` and `norminvgauss` pass. `ncf` passes for very loose tolerances (`5e-5`), and `norminvgauss` passes with looser tolerances still (`5e-4`).

Here's a possibility: several errors occur because of inaccuracies at the tails that we could catch when creating the Hermite spline fails. I could put that part in a `try`/`except` and raise an exception suggesting that the failure is probably due to insufficient accuracy of the distribution and that loosening the `tol` argument might help.

Reply:
Thanks for the precision. I agree that if the errors are in the distributions we should fix the distribution and not spend time here. IMO you are testing this function on a wide range of distributions and I am not sure it would bring more value to add more tests.